src/autoscaler/metricsforwarder/server/custom_metrics_handlers_test.go

```diff
@@ -0,0 +1,306 @@
+					customMetrics := []*models.CustomMetric{
+						&models.CustomMetric{
+							Name: "queuelength", Value: 12, Unit: "unit", InstanceIndex: 1, AppGUID: "an-app-id",
+						},
+					}
+					body, err = json.Marshal(models.MetricsConsumer{InstanceIndex: 0, CustomMetrics: customMetrics})
+					Expect(err).NotTo(HaveOccurred())
+				})
+
+				It("should get the credentials from database and add it to the cache and returns status code 200", func() {
+					Expect(policyDB.GetCustomMetricsCredsCallCount()).To(Equal(1))
+					Expect(resp.Code).To(Equal(http.StatusOK))
+					_, found = credentialCache.Get("an-app-id")
+					Expect(found).To(Equal(true))
+				})
+
+			})
+
+			Context("when credentials neither exists in the cache nor exist in the database", func() {
+				BeforeEach(func() {
+					customMetrics := []*models.CustomMetric{
+						&models.CustomMetric{
+							Name: "queuelength", Value: 12, Unit: "unit", InstanceIndex: 1, AppGUID: "an-app-id",
+						},
+					}
+					body, err = json.Marshal(models.MetricsConsumer{InstanceIndex: 0, CustomMetrics: customMetrics})
+					Expect(err).NotTo(HaveOccurred())
+				})
+
+				It("should search in both cache & database and returns status code 401", func() {
+					Expect(policyDB.GetCustomMetricsCredsCallCount()).To(Equal(1))
+					Expect(resp.Code).To(Equal(http.StatusUnauthorized))
+				})
+
+			})
+
+			Context("when a stale credentials exists in the cache", func() {
+				BeforeEach(func() {
+					credentials.Username = "some-stale-hashed-username"
+					credentials.Password = "some-stale-hashed-password"
+					credentialCache.Set("an-app-id", credentials, 10*time.Minute)
+					scalingPolicy = &models.ScalingPolicy{
+						InstanceMin: 1,
+						InstanceMax: 6,
+						ScalingRules: []*models.ScalingRule{{
+							MetricType:            "queuelength",
+							BreachDurationSeconds: 60,
+							Threshold:             10,
+							Operator:              ">",
+							CoolDownSeconds:       60,
+							Adjustment:            "+1"}}}
+					policyDB.GetAppPolicyReturns(scalingPolicy, nil)
+					policyDB.GetCustomMetricsCredsReturns("$2a$10$YnQNQYcvl/Q2BKtThOKFZ.KB0nTIZwhKr5q1pWTTwC/PUAHsbcpFu", "$2a$10$6nZ73cm7IV26wxRnmm5E1.nbk9G.0a4MrbzBFPChkm5fPftsUwj9G", nil)
+					customMetrics := []*models.CustomMetric{
+						&models.CustomMetric{
+							Name: "queuelength", Value: 12, Unit: "unit", InstanceIndex: 1, AppGUID: "an-app-id",
+						},
+					}
+					body, err = json.Marshal(models.MetricsConsumer{InstanceIndex: 0, CustomMetrics: customMetrics})
+					Expect(err).NotTo(HaveOccurred())
+				})
+
+				It("should search in the database and returns status code 200", func() {
+					Expect(policyDB.GetCustomMetricsCredsCallCount()).To(Equal(1))
+					Expect(resp.Code).To(Equal(http.StatusOK))
+				})
+			})
+		})
+
+		Context("when a request to publish custom metrics comes with malformed request body", func() {
```

Comment:
Can we add more test case for this part?   for example but not limited to:  
1, invalid metrics which conflict with standard metric, 
2 invalid metrics which is not an  allowed one,   
3. allowed metric type set already, but policy is updated , then what will happen ? 
.....
....

Reply:
Added/Modified Unit test cases

Reply:
will you continue to enhance to address  the last suggestion  "allowed metric type set already, but policy is updated , then what will happen? "

Reply:
Added policymanager to keep updating the policy modification.